File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionSum.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.math.BigDecimal;
import java.util.Map;
import java.util.Objects;

public class TransformerFunctionSum extends TransformerFunction {

    public TransformerFunctionSum() {
        super(FunctionDescription.of(
            Map.of(
            "default", ArgumentType.of(ArgType.BigDecimal).position(0).defaultBigDecimal(BigDecimal.ZERO),
            "by", ArgumentType.of(ArgType.Transformer).position(1).defaultIsNull(true)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null || streamer.knownAsEmpty())
            return null;
        var by = context.getJsonElement("by", false);
        var def = Objects.requireNonNullElse(context.getBigDecimal("default"), BigDecimal.ZERO);
        var adapter = context.getAdapter();
        var result = streamer.stream()
                .map(t -> {
                    var res = !adapter.isNull(by) ? context.transformItem(by, t) : t;
                    return adapter.isNull(res) ? def : adapter.getNumberAsBigDecimal(res);
                })
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // cap scale at max
        if (result.scale() > FunctionHelpers.MAX_SCALE) {
            result = result.setScale(FunctionHelpers.MAX_SCALE, FunctionHelpers.MAX_SCALE_ROUNDING);
        }
        return result;
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionSum.ts
import BigNumber from "bignumber.js";
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined } from "../JsonHelpers";
import { BigDecimal, MAX_SCALE, MAX_SCALE_ROUNDING } from "./common/FunctionHelpers";

class TransformerFunctionSum extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        default: { type: ArgType.BigDecimal, position: 0, defaultBigDecimal: 0 },
        by: { type: ArgType.Transformer, position: 1, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getJsonElementStreamer(null);
    if (value == null || value.knownAsEmpty()) {
      return null;
    }
    const hasBy = context.has("by");
    const by = await context.getJsonElement("by", false);
    const _default = (await context.getBigDecimal("default")) ?? BigDecimal(0);
    let size = 0;
    let sum = await value
      .stream()
      .map(async t => {
        size++;
        const res = !hasBy ? t : await context.transformItem(by, t);
        return isNullOrUndefined(res) ? _default : BigDecimal(res);
      })
      .reduce((a: BigNumber, c) => a.plus(c));
    if ((sum.decimalPlaces() ?? 0) > MAX_SCALE) {
      sum = sum.decimalPlaces(MAX_SCALE, MAX_SCALE_ROUNDING);
    }
    return sum;
  }
}

export default TransformerFunctionSum;
